(DOCSP-13197): Stitch deprecation doc - node (#654)
mongodb/docs-realm: source/node/migrate.txt

.. _node-migrating-from-stitch-to-realm:

============================
Upgrade from Stitch to Realm
============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------
If you have an existing app built with the Stitch SDK, you should migrate your
app to use the new Realm SDK.  While much of the application logic and flow of
information hasn't changed, there are a few changes to how it connects to the
{+backend+} backend.

New Features
------------

- The Realm Node SDK now includes {+client-database+}, which includes
  local object storage and the ability to synchronize document changes

>   across devices over the internet.

Author: This page doesn't discuss the Web SDK at all. There is another ticket to discuss the Web SDK: https://jira.mongodb.org/browse/DOCSP-13594

I've also added a note to the top of this page to make this more clear to readers.

Deprecated Features
-------------------

- The {+realm+} SDKs do not provide hooks for calling Services (like the 
  :doc:`Twilio <services/twilio>` and :doc:`AWS <services/aws>` services). We 

>   recommend using the AWS and/or Twilio ``npm`` packages within {+realm+} 

Author: 👍

  functions to connect to these services moving forward. For more information, 
  see :doc:`Upload External Dependencies </functions/upload-external-dependencies>`.

Changes
-------

.. list-table::
   :header-rows: 1
   :widths: 50 50

   * - What's Changed?
     - Solution
   

>    * - The backend app connections previously accessed via `StitchAppClient <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchappclient.html>`__ are now available through :js-sdk:`Realm.App <Realm.App.html>`. Use the ``Realm.App`` class to access authentication functionality such as logging in and getting the current user. The ``Realm.App`` is not an interface to access {+backend+} functions.

Author: It is another bullet actually. It's the last bullet. Is there a way to restructure this sentence to be more readable, or do you think this suffices (with knowing that the last bullet goes further in-depth on this).

>    * - The backend app connections previously accessed via `StitchAppClient <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchappclient.html>`__ are now available through :js-sdk:`Realm.App <Realm.App.html>`. Use the ``Realm.App`` class to access authentication functionality such as logging in and getting the current user. The ``Realm.App`` is not an interface to access {+backend+} functions.

Author: Update: I've omitted this bit entirely in favor of the bullet that's exclusively about functions.

     - Migrate all occurrences of ``StitchAppClient`` to ``Realm.App``.

   * - Application users previously accessed via `StitchUser <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchuser.html>`__ are now accessed through :js-sdk:`Realm.User <Realm.User.html>`.
     - Migrate all occurrences of ``StitchUser`` to ``Realm.User``.

   * - The login API has changed from `stitchAppClient.auth.loginWithCredential() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchauth.html#loginwithcredential>`__ to :js-sdk:`Realm.App.logIn() <Realm.App.html#logIn>`
     - Rewrite all authentication logic to use the :ref:`new login API <node-authenticate>`.

   * - The API to log users out has changed from `stitchAppClient.auth.logout() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchauth.html#logout>`__ to :js-sdk:`Realm.User.logOut() <Realm.User.html#logOut>`.
     - Rewrite the application logic to log out using the ``Realm.User`` object's ``logOut()`` method.

>    * - The Stitch SDK required developers to import authentication credentials from the "mongodb-stitch-server-sdk". The Realm SDK has a class, :js-sdk:`Realm.Credentials <Realm.Credentials.html>` for creating credentials. ``Realm.Credentials`` contains methods to create credentials for each authentication provider.

Author: * "mongodb-stitch-server-sdk" is the sdk that is discussed in this page. It is the equivalent to the Realm Node SDK. Whereas, "mongodb-stitch-browser-sdk" is the equivalent to the Realm Web SDK. I'll add a note at the top of the page to specify.

* The sentence: "The Realm SDK has a class, Realm.Credentials, for creating credentials"  does not contain passive voice. [Passive voice](https://webapps.towson.edu/ows/activepass.htm#:~:text=the%20active%20voice.-,Passive%20voice,by%20the%20verb%20%2D%20or%20passive.) is when an action (verb) is performed on a subject, without describing who the subject is. Even if it was passive voice, I don't think passive voice is bad in the "whats changed" column because it is not an action oriented column (vs "solution" which would be require an active voice).

    Here's a handy tool for detecting passive voice: https://datayze.com/passive-voice-detector (I don't use it, I use Grammarly, but this is a completely free option)

* Good idea! Changed it to your suggestion now

     - Replace imports of authentication credentials with method calls for each respective authentication provider through the ``Realm.Credentials`` class.

>    * - Instead of calling {+backend+} functions using the app client directly via the method `stitchAppClient.callFunction() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js/4/interfaces/stitchappclient.html#callfunction>`__, you must call server {+backend+} functions by using the :js-sdk:`callFunction() <Realm.User.html#callFunction>` method of a logged in ``Realm.User``. 

Reviewer: This bullet point gets a little verbose in the middle. I think it's also a good idea to compare and contrast the approaches by naming the SDKs explicitly, instead of always using implicit "stitch first, then realm" ordering.
```suggestion
   * - Instead of calling {+backend+} functions using the app client directly via the Stitch SDK method `stitchAppClient.callFunction() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js/4/interfaces/stitchappclient.html#callfunction>`__, call {+service-short+} functions in the Realm SDK using the :js-sdk:`callFunction() <Realm.User.html#callFunction>` method of a logged in ``Realm.User``. 
```

Author: Resolving because this is outdated now. Though I took your suggestions (albeit slightly altered)

     - Call {+backend+} functions using ``Realm.User.callFunction()`` instead of an app client.

>    * - The MongoDB Data Access API, which contains objects like :js-sdk:`MongoDatabase <Realm-MongoDBDatabase.html>`, and :js-sdk:`MongoCollection <Realm.MongoDBCollection>` was previously accessed by passing the `RemoteMongoClient.Factory <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/modules/remotemongoclient.html#factory>`__ to the `StitchAppClient.getServiceClient() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchappclient.html#getserviceclient>`__.  It is now accessed via the user object via :js-sdk:`Realm.User.mongoClient <Realm.User.html#mongoClient>`.

Reviewer: The MongoDBCollection link here is broken and leads to a 403 forbidden.

Author: whoops forgot to comment on this one. It's fixed tho

.. _node-migrating-from-stitch-to-realm:

============================
Upgrade from Stitch to Realm
============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------
If you have an existing app built with the Stitch SDK, you should migrate your
app to use the new Realm SDK.  While much of the application logic and flow of
information hasn't changed, there are a few changes to how it connects to the
{+realm+} backend.

.. note::

>    The following page presents the differences between the Realm JS SDK, used to

Author: ```suggestion
   The following page presents the differences between the Realm JS SDK, which you can use to 
```

     - Rewrite the application logic to log out using the ``Realm.User`` object's ``logOut()`` method.

   * - The Stitch SDK required developers to import authentication credentials from the "mongodb-stitch-server-sdk". The Realm SDK has a class, :js-sdk:`Realm.Credentials <Realm.Credentials.html>`, that creates credentials for each authentication provider.
     - Replace imports of authentication credentials with method calls for each respective authentication provider through the ``Realm.Credentials`` class.

   * - Instead of calling {+backend+} functions using the app client directly via the method `stitchAppClient.callFunction() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js/4/interfaces/stitchappclient.html#callfunction>`__, you must call server {+backend+} functions using the :js-sdk:`callFunction() <Realm.User.html#callFunction>` method of a logged-in ``Realm.User``. 
     - Call {+backend+} functions using ``Realm.User.callFunction()`` instead of an app client.

>   * - The interface to the MongoDB Data Access API, containing objects like :js-sdk:`MongoDatabase <Realm-MongoDBDatabase.html>`, and :js-sdk:`MongoCollection <Realm.MongoDBCollection>`, has changed. Through Stitch, passing the `RemoteMongoClient.Factory <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/modules/remotemongoclient.html#factory>`__ and the string "mongodb-atlas" to `StitchAppClient.getServiceClient() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchappclient.html#getserviceclient>`__ would create a remote Mongodb client. The Realm JS SDK creates a remote MongoDB client through the :js-sdk:`Realm.User.mongoClient <Realm.User.html#mongoClient>` method on the user object.

Author: ```suggestion
   * - The interface to the MongoDB Data Access API, containing objects like :js-sdk:`MongoDatabase <Realm-MongoDBDatabase.html>`, and :js-sdk:`MongoCollection <Realm.MongoDBCollection.html>`, has changed. Through Stitch, passing the `RemoteMongoClient.Factory <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/modules/remotemongoclient.html#factory>`__ and the string "mongodb-atlas" to `StitchAppClient.getServiceClient() <http://stitch-sdks.s3-website-us-east-1.amazonaws.com/stitch-sdks/js-server/4/interfaces/stitchappclient.html#getserviceclient>`__ would create a remote Mongodb client. The Realm JS SDK creates a remote MongoDB client through the :js-sdk:`Realm.User.mongoClient <Realm.User.html#mongoClient>` method on the user object.
```

>     - Replace all ``mongoClient`` instatiations with ``Realm.User.mongoClient``.

Author: ```suggestion
    - Replace all ``mongoClient`` instantiations with ``Realm.User.mongoClient``.
```

>     - Replace all ``mongoClient`` instantiations with ``Realm.User.mongoClient``.

Author: ```suggestion
     - Replace all ``mongoClient`` instantiations with ``Realm.User.mongoClient``.
```
